homeassistant/components/obihai/__init__.py

```diff
@@ -1,13 +1,26 @@
 """The Obihai integration."""
 
 from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_USERNAME
 from homeassistant.core import HomeAssistant
 
+from .connectivity import ObihaiConnection
 from .const import PLATFORMS
 
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up from a config entry."""
+    requester = ObihaiConnection(
+        entry.data[CONF_HOST],
+        username=entry.data[CONF_USERNAME],
+        password=entry.data[CONF_PASSWORD],
+    )
+    await hass.async_add_executor_job(requester.update)
+
+    new_unique_id = await hass.async_add_executor_job(requester.pyobihai.get_device_mac)
+    if entry.unique_id != new_unique_id:
+        hass.config_entries.async_update_entry(entry, unique_id=new_unique_id)
+
```

Comment:
Instead of checking this at every startup, we could bump `ObihaiFlowHandler.VERSION` and migrate the `unique_id` to MAC in `async def async_migrate(self, hass: HomeAssistant) -> bool:` which will be called if the `VERSION` is not up to date.